lib/recommendation.js

```diff
@@ -0,0 +1,61 @@
+var XMLHttpRequest = Cc['@mozilla.org/xmlextras/xmlhttprequest;1'].createInstance();
+
+
+class RecommendationServer {
+  /*
+  Class providing a connection to a recommendation server and a method for
+  querying it. This class should be initialized once and exported as a
+  singleton to modules that will use it.
+
+  Usage:
+
+  let recommendation = new RecommendationServer('http://universal-search.dev');
+
+  */
+  constructor(server) {
+    this.server = server;
+  }
+
+  _url(query) {
+    return `${this.server}/?q=${encodeURIComponent(query)}`;
+  }
+
+  make(query) {
+    /*
+    Method to query a recommendation server. Returns a promise that resolves to
+    a recommendation or rejects to a reason that a recommendation is not
+    available.
+
+    Usage:
+
+    recommendation.make('hello wor').then(data => {
+      // Recommendation returned as `data` object literal.
+    }, reason => {
+      // No recommendation returned for `reason` string.
+    });
+
+    */
+    const url = this._url(query);
+    return new Promise((resolve, reject) => {
+      try {
+        var req = new XMLHttpRequest();
```

Comment:
This library should keep a single xhr pointer and abort in-flight requests when new searches begin